src/System.Private.CoreLib/shared/System/IO/BinaryReader.cs

```diff
@@ -260,94 +257,45 @@ public virtual char ReadChar()
             return (char)value;
         }
 
-        public virtual short ReadInt16()
-        {
-            FillBuffer(2);
-            return (short)(_buffer[0] | _buffer[1] << 8);
-        }
+        public virtual short ReadInt16() => BinaryPrimitives.ReadInt16LittleEndian(InternalRead(2));
 
         [CLSCompliant(false)]
-        public virtual ushort ReadUInt16()
-        {
-            FillBuffer(2);
-            return (ushort)(_buffer[0] | _buffer[1] << 8);
-        }
-
-        public virtual int ReadInt32()
-        {
-            if (_isMemoryStream)
-            {
-                if (_stream == null)
-                {
-                    throw Error.GetFileNotOpen();
-                }
-
-                // read directly from MemoryStream buffer
-                MemoryStream mStream = _stream as MemoryStream;
-                Debug.Assert(mStream != null, "_stream as MemoryStream != null");
-
-                return mStream.InternalReadInt32();
-            }
-            else
-            {
-                FillBuffer(4);
-                return (int)(_buffer[0] | _buffer[1] << 8 | _buffer[2] << 16 | _buffer[3] << 24);
-            }
-        }
+        public virtual ushort ReadUInt16() => BinaryPrimitives.ReadUInt16LittleEndian(InternalRead(2));
 
+        public virtual int ReadInt32() => BinaryPrimitives.ReadInt32LittleEndian(InternalRead(4));
         [CLSCompliant(false)]
-        public virtual uint ReadUInt32()
-        {
-            FillBuffer(4);
-            return (uint)(_buffer[0] | _buffer[1] << 8 | _buffer[2] << 16 | _buffer[3] << 24);
-        }
-
-        public virtual long ReadInt64()
-        {
-            FillBuffer(8);
-            uint lo = (uint)(_buffer[0] | _buffer[1] << 8 |
-                             _buffer[2] << 16 | _buffer[3] << 24);
-            uint hi = (uint)(_buffer[4] | _buffer[5] << 8 |
-                             _buffer[6] << 16 | _buffer[7] << 24);
-            return (long)((ulong)hi) << 32 | lo;
-        }
-
+        public virtual uint ReadUInt32() => BinaryPrimitives.ReadUInt32LittleEndian(InternalRead(4));
+        public virtual long ReadInt64() => BinaryPrimitives.ReadInt64LittleEndian(InternalRead(8));
         [CLSCompliant(false)]
-        public virtual ulong ReadUInt64()
-        {
-            FillBuffer(8);
-            uint lo = (uint)(_buffer[0] | _buffer[1] << 8 |
-                             _buffer[2] << 16 | _buffer[3] << 24);
-            uint hi = (uint)(_buffer[4] | _buffer[5] << 8 |
-                             _buffer[6] << 16 | _buffer[7] << 24);
-            return ((ulong)hi) << 32 | lo;
-        }
+        public virtual ulong ReadUInt64() => BinaryPrimitives.ReadUInt64LittleEndian(InternalRead(8));
+
+        public virtual Guid ReadGuid() => new Guid(InternalRead(16));
```

Comment:
There was no ReadGuid method before. Why are you adding one?

Reply:
This turned out to be the biggest missing thing in the reader. As only the reader has access to the performant way of reading MemoryStream I think it is worthwhile adding it. If this is ok, I will also add it to the BinaryWriter for symmetry reasons.
What I would think is great improvement is to expose my new memoryStream.InternalReadSpan as public and than other implementors can gain access to speed reading a stream. The thing I wasn't sure about is why MemoryStream has the publiclyVisible property and should we block that function from returning a span when that thing is false. Can you comment about why publiclyVisible is needed and if exposing direct ReadOnlySpan is good idea?

Reply:
The process for adding new public APIs to .NET Core is documented here: https://github.com/dotnet/corefx/blob/master/Documentation/project-docs/api-review-process.md . If you think there should be more APIs on BinaryReader, start with filing an issue with the proposal in corefx repo.